athena/lib/larv/larv_set_busy.3
.\" $Id: larv_set_busy.3,v 1.1.2.1 1998-09-24 13:54:50 ghudson Exp $
.\"
.\" Copyright 1998 by the Massachusetts Institute of Technology.
.\"
.\" Permission to use, copy, modify, and distribute this
.\" software and its documentation for any purpose and without
.\" fee is hereby granted, provided that the above copyright
.\" notice appear in all copies and that both that copyright
.\" notice and this permission notice appear in supporting
.\" documentation, and that the name of M.I.T. not be used in
.\" advertising or publicity pertaining to distribution of the
.\" software without specific, written prior permission.
.\" M.I.T. makes no representations about the suitability of
.\" this software for any purpose.  It is provided "as is"
.\" without express or implied warranty.
.TH LARV_SET_BUSY 3 "23 Aug 1998"
.SH NAME
larv_set_busy \- Set and report the workstation's busy state
.SH SYNOPSIS
.B int larv_set_busy(int \fIbusy\fP)
.SH DESCRIPTION
The
.B larv_set_busy
function sets the workstation's busy status by creating or deleting
the file
.BR /var/athena/busy .
It also reports the workstation's busy status to each server listed
under the Hesiod name
.B larvnet
with type
.BR sloc .
.SH FILES
.B /etc/athena/rc.conf
.br
.B /var/athena/busy
.SH "SEE ALSO"
.BR busyd (8)
.SH AUTHOR
Greg Hudson, MIT Information Systems
.br
Copyright 1998 by the Massachusetts Institute of Technology.


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


athena/lib/larv/larv_set_busy.c
/* Copyright 1998 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

static const char rcsid[] = "$Id: larv_set_busy.c,v 1.1.2.1 1998-09-24 13:54:51 ghudson Exp $";

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <netdb.h>
#include <hesiod.h>
#include "larvnet.h"
#include "larv.h"
#include "larv_private.h"

int larv_set_busy(int busy)
{
  struct servent *serv;
  struct hostent *host;
  struct sockaddr_in sin;
  unsigned short port;
  int fd, s, len;
  void *ctx;
  char **vec, **vp, buf[LARVNET_MAX_PACKET];

  /* Create or unlink the busy file, as appropriate. */
  if (busy)
    {
      fd = open(LARVNET_PATH_BUSY, O_CREAT, S_IRWXU);
      if (fd < 0)
	return -1;
      else
	close(fd);
    }
  else
    {
      if (unlink(LARVNET_PATH_BUSY) < 0 && errno != ENOENT)
	return -1;
    }

  /* Compose the status packet.  It's important that we do this step
   * after creating or unlinking the busy file, since
   * larv__compose_packet() will check whether the busy file exists.
   */
  len = larv__compose_packet(buf);
  if (len == -1)
    return -1;

  serv = getservbyname("larvnet", "udp");
  port = (serv) ? serv->s_port : htons(LARVNET_FALLBACK_PORT);

  s = socket(AF_INET, SOCK_DGRAM, 0);
  if (s == -1)
    return -1;

  if (hesiod_init(&ctx) != 0)
    {
      close(s);
      return -1;
    }

  vec = hesiod_resolve(ctx, "larvnet", "sloc");
  if (!vec)
    {
      hesiod_end(ctx);
      close(s);
      return (errno == ENOENT) ? 0 : -1;
    }

  /* Send the status packet to each cview service location record. */
  for (vp = vec; *vp; vp++)
    {
      host = gethostbyname(*vp);
      if (!host || host->h_addrtype != AF_INET
	  || host->h_length != sizeof(sin.sin_addr))
	continue;
      memset(&sin, 0, sizeof(sin));
      sin.sin_family = AF_INET;
      sin.sin_port = port;
      memcpy(&sin.sin_addr, host->h_addr, sizeof(sin.sin_addr));
      sendto(s, buf, len, 0, (struct sockaddr *) &sin, sizeof(sin));
    }

  hesiod_free_list(ctx, vec);
  hesiod_end(ctx);
  close(s);
  return 0;
}
